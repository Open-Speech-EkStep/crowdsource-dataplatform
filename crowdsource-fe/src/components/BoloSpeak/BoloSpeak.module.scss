.root {
  z-index: 1;
}

.progress {
  height: 0.625rem;
}

.audioLine {
  background-color: var(--link-color);
  border-radius: 2px;
  height: 5px;
}

.mic {
  box-shadow: 0 0 15px 2px #bdc3c7;
  height: 4rem;
  width: 4rem;
}

.audioWrapper,
.visualizer {
  height: 7.5rem;
}

.erroMsg {
  bottom: 0;
}

.audioPermissionDeniedHeader {
  :global(.btn-close) {
    position: static;
  }
}
