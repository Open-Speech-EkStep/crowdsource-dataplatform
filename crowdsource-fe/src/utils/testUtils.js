import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
