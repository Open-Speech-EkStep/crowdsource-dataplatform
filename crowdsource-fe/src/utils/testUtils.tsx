/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
