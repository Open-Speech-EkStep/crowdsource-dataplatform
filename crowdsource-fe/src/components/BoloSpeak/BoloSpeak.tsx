import React, { Fragment, useEffect, useRef, useState } from 'react';

import getBlobDuration from 'get-blob-duration';
import { Trans, useTranslation } from 'next-i18next';
import Image from 'next/image';
import { useRouter } from 'next/router';
import Container from 'react-bootstrap/Container';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Spinner from 'react-bootstrap/Spinner';

import ButtonControls from 'components/ButtonControls';
import ErrorPopup from 'components/ErrorPopup';
import FunctionalHeader from 'components/FunctionalHeader';
import NoDataFound from 'components/NoDataFound';
import apiPaths from 'constants/apiPaths';
import { AUDIO } from 'constants/Audio';
import {
  INITIATIVE_ACTIONS,
  INITIATIVES_MAPPING,
  INITIATIVES_MEDIA_MAPPING,
} from 'constants/initiativeConstants';
import localStorageConstants from 'constants/localStorageConstants';
import routePaths from 'constants/routePaths';
import { useSubmit } from 'hooks/useFetch';
import useLocalStorage from 'hooks/useLocalStorage';
import useFetch from 'hooks/usePostFetch';
import type { ActionStoreBoloInterface } from 'types/ActionRequestData';
import type { LocationInfo } from 'types/LocationInfo';
import type SpeakerDetails from 'types/SpeakerDetails';
import { getBrowserInfo, getDeviceInfo, getErrorMsg, visualize } from 'utils/utils';

import QuickTips from '../QuickTips';

import styles from './BoloSpeak.module.scss';

interface ResultType {
  data: any;
}

const BoloSpeak = () => {
  const { t } = useTranslation();

  const [contributionLanguage] = useLocalStorage<string>(localStorageConstants.contributionLanguage);

  const [speakerDetails] = useLocalStorage<SpeakerDetails>(localStorageConstants.speakerDetails);
  const [lastSpeakerDetail, setLastSpeakerDetails] = useLocalStorage<SpeakerDetails | null>(
    localStorageConstants.lastSpeakerDetails
  );
  const [locationInfo] = useLocalStorage<LocationInfo>(localStorageConstants.locationInfo);

  const [showPlayButton] = useState(false);

  const [showStartRecording, setShowStartRecording] = useState(true);
  const [showStopRecording, setShowStopRecording] = useState(false);
  const [showReRecording, setShowReRecording] = useState(false);
  const [showWarningmsg, setShowWarningMsg] = useState(false);
  const [showQuickTips, setShowQuickTips] = useState(false);
  const [showAudioController, setShowAudioController] = useState(false);
  const [audioError, setShowAudioError] = useState(false);
  const [contributionData, setContributionData] = useState([]);
  const [duration, setDuration] = useState(0);

  const [remainingSec, setRemainingSec] = useState(AUDIO.WARNING_COUNT_START);

  const [recordedAudio, setRecordedAudio] = useState<string | undefined>();

  const [currentDataIndex, setCurrentDataIndex] = useState<number>(0);
  const router = useRouter();
  const { locale: currentLocale } = useRouter();
  const [showUIData, setShowUIdata] = useState({
    media_data: '',
    dataset_row_id: '0',
  });

  const [gumStream, setGumStream] = useState<MediaStream>();

  const [timerTimeoutKey, setTimerTimeoutKey] = useState<any>();

  const [clearTimeoutKey, setClearTimeoutKey] = useState<any>();

  const audioController = useRef<any>();
  const mediaRecorder = useRef<any>();
  const audioCtx = useRef<any>();

  let input: any;
  let chunks: any = [];

  const { submit, error: submitError } = useSubmit(apiPaths.store, false);
  const audioEl: any = useRef<HTMLAudioElement>();
  const audio = audioEl.current;
  const { submit: submitSkip, error: skipError } = useSubmit(apiPaths.skip);
  const [showErrorModal, setShowErrorModal] = useState(false);

  const [formData, setFormData] = useState<ActionStoreBoloInterface>({
    speakerDetails: '',
    language: contributionLanguage ?? '',
    type: INITIATIVES_MEDIA_MAPPING.bolo,
    sentenceId: 0,
    state: '',
    country: '',
    device: getDeviceInfo(),
    browser: getBrowserInfo(),
    audioDuration: 0,
    audio_data: null,
  });

  const { data: result, error } = useFetch<ResultType>({
    url: apiPaths.mediaText,
    init: contributionLanguage
      ? {
          body: JSON.stringify({
            language: contributionLanguage,
            userName: speakerDetails?.userName,
          }),
          method: 'POST',
          credentials: 'include',
          mode: 'cors',
          headers: { 'Content-Type': 'application/json' },
        }
      : undefined,
  });

  useEffect(() => {
    if (skipError || submitError || error) {
      setShowErrorModal(true);
    }
  }, [skipError, submitError, error]);

  useEffect(() => {
    if (result && result.data) {
      if (lastSpeakerDetail?.userName !== speakerDetails?.userName) {
        setShowQuickTips(true);
      }
      setLastSpeakerDetails(speakerDetails);
    }
  }, [result, lastSpeakerDetail, speakerDetails, setLastSpeakerDetails]);

  useEffect(() => {
    if (result && result.data) {
      setContributionData(result.data);
      setShowUIdata(result.data[currentDataIndex]);
    }
  }, [currentDataIndex, result]);

  const setDataCurrentIndex = (index: number) => {
    if (index === contributionData.length - 1) {
      router.push(`/${currentLocale}${routePaths.boloIndiaContributeThankYou}`, undefined, {
        locale: currentLocale,
      });
    } else {
      setCurrentDataIndex(index + 1);
      setShowUIdata(contributionData[index + 1]);
    }
  };

  const resetState = () => {
    setRecordedAudio('');
    setDuration(0);
    setShowStartRecording(true);
    setShowStopRecording(false);
    setShowReRecording(false);
    setShowAudioController(false);
    audioController?.current?.classList.add('d-none');
  };

  const setRemainingCount = (value: number) => {
    if (value) {
      const interval = setInterval(() => {
        setRemainingSec(value);
        value--;
        if (value < 0) {
          clearInterval(interval);
        }
      }, 1000);
    }
  };

  const startTimer = () => {
    setTimerTimeoutKey(
      setTimeout(() => {
        setShowWarningMsg(true);
        setRemainingCount(AUDIO.WARNING_COUNT_START - 1);
      }, (AUDIO.MAX_DURATION - 5) * 1000)
    );

    setClearTimeoutKey(
      setTimeout(() => {
        setShowWarningMsg(false);
        onStopRecording();
      }, AUDIO.MAX_DURATION * 1000)
    );
  };

  /* istanbul ignore next */
  const onRecordingStop = async () => {
    audioController?.current?.classList.remove('d-none');
    setShowAudioController(true);
    const blob = new Blob(chunks, { type: 'audio/wav' });
    chunks = [];
    setFormData({
      ...formData,
      audio_data: blob,
    });
    const audioURL = URL.createObjectURL(blob);
    setRecordedAudio(audioURL);
    audio.onloadedmetadata = async () => {
      // it should already be available here
      // handle chrome's bug
      if (audio.duration === Infinity) {
        // set it to bigger than the actual duration
        audio.currentTime = 1e101;
        audio.ontimeupdate = () => {
          audio.ontimeupdate = () => {};
          // setting player currentTime back to 0 can be buggy too, set it first to .1 sec
          audio.currentTime = 0.1;
          audio.currentTime = 0;
        };
      }
      const duration = await getBlobDuration(blob);
      validateAudioDuration(duration);
    };
  };

  const recordAudio = async () => {
    try {
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      setGumStream(stream);
      const AudioContext = window.AudioContext;
      if (audioCtx && audioCtx.current) {
        audioCtx.current.close();
      }
      audioCtx.current = new AudioContext();
      const audioAnalyser = audioCtx.current.createAnalyser();
      //new audio context to help us record
      input = audioCtx.current.createMediaStreamSource(stream);
      input.connect(audioAnalyser);
      // eslint-disable-next-line no-undef
      const visualizer: any = document.getElementById('visualizer');
      visualize(visualizer, audioAnalyser);
      /* Create the Recorder object and configure to record mono sound (1 channel) Recording 2 channels will double the file size */
      mediaRecorder.current = new MediaRecorder(stream);

      //start the recording process
      mediaRecorder.current.start();

      //automatically click stop button after 20 seconds
      mediaRecorder.current.ondataavailable = (e: any) => {
        chunks.push(e?.data);
      };

      mediaRecorder.current.onstop = () => {
        onRecordingStop();
      };
      startTimer();
    } catch (e) {
      console.log(e);
    }
  };

  const validateAudioDuration = (duration: number) => {
    if (duration < AUDIO.MIN_DURATION) {
      setShowAudioError(true);
    } else {
      setShowAudioError(false);
    }
    setDuration(duration);
  };

  const stopAudio = () => {
    setShowWarningMsg(false);
    clearTimeout(clearTimeoutKey);
    clearTimeout(timerTimeoutKey);
    mediaRecorder?.current.stop();
    gumStream?.getAudioTracks()[0].stop();
  };

  const onStartRecording = () => {
    setShowStartRecording(false);
    setShowStopRecording(true);
    recordAudio();
  };

  const onStopRecording = () => {
    setShowReRecording(true);
    setShowStopRecording(false);
    stopAudio();
  };

  const onRerecord = () => {
    setRecordedAudio('');
    audioController?.current?.classList.add('d-none');
    setShowStopRecording(true);
    setShowAudioError(false);
    setShowReRecording(false);
    setShowAudioController(false);
    recordAudio();
  };

  /* istanbul ignore next */
  const onSubmitContribution = async () => {
    setDataCurrentIndex(currentDataIndex);
    resetState();
    const fd = new FormData();
    fd.append('language', `${contributionLanguage}`);
    fd.append('sentenceId', showUIData.dataset_row_id);
    fd.append('country', `${locationInfo?.country}`);
    fd.append('state', `${locationInfo?.regionName}`);
    fd.append('audioDuration', `${duration}`);
    fd.append(
      'speakerDetails',
      JSON.stringify({
        userName: speakerDetails?.userName,
        age: speakerDetails?.age,
        motherTongue: speakerDetails?.motherTongue,
        gender: speakerDetails?.gender,
      })
    );
    fd.append('audio_data', formData.audio_data);
    fd.append('device', getDeviceInfo());
    fd.append('browser', getBrowserInfo());
    fd.append('type', INITIATIVES_MEDIA_MAPPING.bolo);

    submit(fd);
  };

  const onSkipContribution = () => {
    setDataCurrentIndex(currentDataIndex);
    resetState();
    submitSkip(
      JSON.stringify({
        device: getDeviceInfo(),
        browser: getBrowserInfo(),
        userName: speakerDetails?.userName,
        language: contributionLanguage,
        sentenceId: showUIData.dataset_row_id,
        state_region: locationInfo?.regionName,
        country: locationInfo?.country,
        type: INITIATIVES_MEDIA_MAPPING.bolo,
      })
    );
  };

  if (!result && !error) {
    return <Spinner data-testid="StatsSpinner" animation="border" variant="light" />;
  }

  return (
    <Fragment>
      {contributionData && result?.data?.length !== 0 ? (
        <Fragment>
          <QuickTips showQuickTips={showQuickTips} />
          <div className="pt-4 px-2 px-lg-0 pb-8">
            <FunctionalHeader
              onSuccess={onSkipContribution}
              type="contribute"
              initiative={INITIATIVES_MAPPING.bolo}
              action={INITIATIVE_ACTIONS[INITIATIVES_MAPPING.bolo]['contribute']}
              showMic={true}
            />
            <Container fluid="lg" className="mt-5">
              <div data-testid="BoloSpeak" className={`${styles.root}`}>
                <div
                  className={`d-flex justify-content-center align-items-center my-9 mt-md-12 mb-md-10 text-center display-1`}
                >
                  {showUIData?.media_data}
                </div>
                <div className={`${styles.audioWrapper} d-flex flex-column justify-content-center`}>
                  <div
                    ref={audioController}
                    className="position-relative d-flex d-none flex-column align-items-center text-center pb-8"
                  >
                    {/* eslint-disable-next-line jsx-a11y/media-has-caption */}
                    <audio
                      ref={audioEl}
                      data-testid="boloAudioElement"
                      controls
                      className="d-flex shadow-grey rounded-24"
                      tabIndex={-1}
                      src={recordedAudio}
                    ></audio>
                    {audioError && (
                      <div
                        className={`${styles.erroMsg} position-absolute d-flex text-danger justify-content-center align-items-center mt-md-2 text-center display-6`}
                      >
                        <span> {t('audioValidationMessage')} </span>
                      </div>
                    )}
                  </div>
                  {!showAudioController && !showStartRecording && (
                    <div className="d-flex align-items-center">
                      <div
                        className={`${styles.mic} d-flex justify-content-center align-items-center rounded-50 bg-danger`}
                      >
                        <Image src="/images/mic_white.svg" width="40" height="40" alt="Mic Icon" />
                      </div>
                      <canvas id="visualizer" className={`${styles.visualizer} ms-5 flex-fill`} />
                    </div>
                  )}
                  {showStartRecording && <span className={`${styles.audioLine} d-flex w-100`} />}
                </div>
                {showWarningmsg && (
                  <div
                    className={`d-flex justify-content-center align-items-center my-9 mt-md-9 mb-md-5 text-center display-5`}
                  >
                    <Trans
                      i18nKey="remainingAudioTextDurationWarning"
                      defaults="remainingAudioTextDurationWarning"
                      values={{
                        remainingSec: remainingSec,
                      }}
                    />
                  </div>
                )}
                <div className="mt-12 mt-md-14">
                  <ButtonControls
                    playButton={showPlayButton}
                    submitDisable={!recordedAudio || duration < 2}
                    onSubmit={onSubmitContribution}
                    onSkip={onSkipContribution}
                    cancelButton={false}
                    startRecordingButton={showStartRecording}
                    stopRecordingButton={showStopRecording}
                    reRecordButton={showReRecording}
                    onStart={onStartRecording}
                    onStop={onStopRecording}
                    onRerecord={onRerecord}
                  />
                </div>

                <div className="d-flex align-items-center mt-10 mt-md-14">
                  <div className="flex-grow-1">
                    <ProgressBar
                      now={(currentDataIndex + 1) * (100 / contributionData.length)}
                      variant="primary"
                      className={styles.progress}
                    />
                  </div>
                  <span className="ms-5">
                    {currentDataIndex + 1}/{contributionData.length}
                  </span>
                </div>
              </div>
            </Container>
          </div>
        </Fragment>
      ) : (
        <div className="d-flex flex-grow-1 align-items-center">
          <NoDataFound
            url={routePaths.boloIndiaHome}
            title={t('textContributeNoDataThankYouMessage')}
            text={t('noDataMessage', { language: t(`${contributionLanguage}`) })}
            buttonLabel={t('textBackToInitiativePrompt')}
          />
        </div>
      )}
      {(skipError || error || submitError) && (
        <ErrorPopup
          show={showErrorModal}
          errorMsg={getErrorMsg(skipError || error || submitError)}
          onHide={() => setShowErrorModal(false)}
        />
      )}
    </Fragment>
  );
};

export default BoloSpeak;
