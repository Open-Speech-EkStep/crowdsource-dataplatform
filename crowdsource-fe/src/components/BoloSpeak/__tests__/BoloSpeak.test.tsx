import { when } from 'jest-when';

import { render, screen, userEvent, waitFor } from 'utils/testUtils';

import BoloSpeak from '../BoloSpeak';

describe('BoloSpeak', () => {
  const resultData = {
    data: [
      {
        dataset_row_id: 371765,
        media_data: 'बल्कि मजबूरी थी 20',
        source_info: null,
      },
      {
        dataset_row_id: 371766,
        media_data: 'बल्कि मजबूरी थी 21',
        source_info: null,
      },
      {
        dataset_row_id: 371767,
        media_data: 'बल्कि मजबूरी थी 22',
        source_info: null,
      },
      {
        dataset_row_id: 371768,
        media_data: 'बल्कि मजबूरी थी 23',
        source_info: null,
      },
      {
        dataset_row_id: 371769,
        media_data: 'बल्कि मजबूरी थी 24',
        source_info: null,
      },
    ],
  };
  const setup = async (resultData: any) => {
    when(localStorage.getItem)
      .calledWith('contributionLanguage')
      .mockImplementation(() => 'Hindi');

    const locationInfo = {
      country: 'India',
      regionName: 'National Capital Territory of Delhi',
    };

    when(localStorage.getItem)
      .calledWith('locationInfo')
      .mockImplementation(() => JSON.stringify(locationInfo));

    const speakerDetails = {
      userName: 'abc',
      motherTongue: '',
      age: '',
      gender: '',
      language: 'Hindi',
      toLanguage: '',
    };

    when(localStorage.getItem)
      .calledWith('speakerDetails')
      .mockImplementation(() => JSON.stringify(speakerDetails));

    fetchMock.doMockOnceIf('/media/text').mockResponseOnce(JSON.stringify(resultData));
    const renderResult = render(<BoloSpeak />);

    await waitFor(() => {
      expect(fetchMock).toBeCalledWith('/media/text', {
        body: JSON.stringify({
          language: 'Hindi',
          userName: 'abc',
        }),
        headers: { 'Content-Type': 'application/json' },
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
      });
    });

    return renderResult;
  };

  it('should start the recording', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    expect(screen.getByRole('button', { name: 'startRecording' })).toBeEnabled();
    userEvent.click(screen.getByRole('button', { name: 'startRecording' }));
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'stopRecording' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });
  });

  it('should show the thank you message when no data present', async () => {
    const result = { data: [] };
    await setup(result);

    await waitFor(() => {
      expect(screen.getByText('textContributeNoDataThankYouMessage')).toBeInTheDocument();
    });
  });

  it('stop button should stop the audio recording', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    expect(screen.getByRole('button', { name: 'startRecording' })).toBeEnabled();
    userEvent.click(screen.getByRole('button', { name: 'startRecording' }));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'stopRecording' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });

    userEvent.click(screen.getByRole('button', { name: 'stopRecording' }));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'reRecord' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });
  });

  it('re-record button should start recording the audio again', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    expect(screen.getByRole('button', { name: 'startRecording' })).toBeEnabled();
    userEvent.click(screen.getByRole('button', { name: 'startRecording' }));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'stopRecording' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });

    userEvent.click(screen.getByRole('button', { name: 'stopRecording' }));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'reRecord' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });

    userEvent.click(screen.getByRole('button', { name: 'reRecord' }));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'stopRecording' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'submit' })).toBeDisabled();
    });
  });

  it('should test the skip functionality', async () => {
    const url = '/skip';
    const successResponse = { message: 'Skipped successfully.', statusCode: 200 };

    await setup(resultData);
    fetchMock.doMockOnceIf(url).mockResponseOnce(JSON.stringify(successResponse));

    expect(screen.getByRole('button', { name: 'skip' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    await waitFor(() => {
      expect(fetchMock).toBeCalledWith(url, {
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          device: 'android 11',
          browser: 'Chrome 13',
          userName: 'abc',
          language: 'Hindi',
          sentenceId: 371765,
          state_region: 'National Capital Territory of Delhi',
          country: 'India',
          type: 'text',
        }),
      });
    });
  });

  it('should show the error popup when api throw the error and close modal on clicking button', async () => {
    const url = '/skip';
    const errorResponse = new Error('Some error');
    await setup(resultData);
    fetchMock.doMockOnceIf(url).mockRejectOnce(errorResponse);
    expect(screen.getByRole('button', { name: 'skip' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'skip' }));
    await waitFor(() => {
      expect(fetchMock).toBeCalledWith(url, {
        method: 'POST',
        credentials: 'include',
        mode: 'cors',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          device: 'android 11',
          browser: 'Chrome 13',
          userName: 'abc',
          language: 'Hindi',
          sentenceId: 371765,
          state_region: 'National Capital Territory of Delhi',
          country: 'India',
          type: 'text',
        }),
      });
    });
    await waitFor(() => {
      expect(screen.getByText('apiFailureError')).toBeInTheDocument();
    });

    await waitFor(() => {
      userEvent.click(screen.getByRole('button', { name: 'close' }));
    });

    await waitFor(() => {
      expect(screen.queryByText('apiFailureError')).not.toBeInTheDocument();
    });
  });

  it('should go to thank you page after 5 skip sentences', async () => {
    await setup(resultData);

    expect(screen.getByRole('button', { name: 'skip' })).toBeEnabled();

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));

    userEvent.click(screen.getByRole('button', { name: 'skip' }));
  });
});
