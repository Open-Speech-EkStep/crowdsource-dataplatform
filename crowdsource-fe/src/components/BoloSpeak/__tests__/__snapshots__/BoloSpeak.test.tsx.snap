// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoloSpeak should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="pt-4 px-2 px-lg-0 pb-8"
  >
    <header>
      <div
        class="d-flex justify-content-between align-items-center px-3 px-md-6"
      >
        <div
          class="root d-flex align-items-center"
          data-testid="Breadcrumbs"
        >
          <a
            class="link"
            href="/bolo-india"
          >
            <div
              class="d-flex align-items-center cursor-pointer"
            >
              <div
                class="icon d-flex"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNDciIGhlaWdodD0iNDIiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <img
                    alt="boloLogo"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="/_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-boloIndiaLogo.svg&w=96&q=75"
                    srcset="/_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-boloIndiaLogo.svg&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fen%2Flogos%2Fen-boloIndiaLogo.svg&w=96&q=75 2x"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </div>
              </div>
              <span
                class="initiative d-none d-md-block ms-1 ms-md-3 font-family-rowdies"
              >
                bolo india
              </span>
            </div>
          </a>
          <span
            class="mx-1 mx-md-3 text-primary-40"
          >
            /
          </span>
          <span
            class="display-5"
          >
            speak
          </span>
        </div>
        <div
          class="d-flex"
        >
          <div>
            <button
              class="root rounded d-flex align-items-center px-3 text-primary text-decoration-none info normal btn btn-link"
              data-testid="reportIconAlt_button"
              type="button"
            >
              <div
                style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
              >
                <div
                  style="box-sizing: border-box; display: block; max-width: 100%;"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                    style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                  />
                </div>
                <img
                  alt="reportIconAlt"
                  data-nimg="intrinsic"
                  decoding="async"
                  src="/_next/image?url=%2Fimages%2Freport.svg&w=48&q=75"
                  srcset="/_next/image?url=%2Fimages%2Freport.svg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Freport.svg&w=48&q=75 2x"
                  style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                />
                <noscript />
              </div>
              <span
                class="text ms-2 d-none d-md-block"
              >
                report
              </span>
            </button>
          </div>
          <div
            class="ms-2 ms-md-4"
          >
            <div
              class="position-relative"
            >
              <button
                class="root rounded d-flex align-items-center px-3 text-primary text-decoration-none info normal btn btn-link"
                data-testid="testYourSpeaker_button"
                type="button"
              >
                <div
                  style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
                >
                  <div
                    style="box-sizing: border-box; display: block; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                      style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
                    />
                  </div>
                  <img
                    alt="testYourSpeaker"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="/_next/image?url=%2Fimages%2Fspeaker.svg&w=48&q=75"
                    srcset="/_next/image?url=%2Fimages%2Fspeaker.svg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fspeaker.svg&w=48&q=75 2x"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </div>
                <span
                  class="text ms-2 d-md-none"
                >
                  test
                </span>
                <span
                  class="text ms-2 d-none d-md-block"
                >
                  testYourSpeaker
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="mt-5 container-lg"
    >
      <div
        class="root"
        data-testid="BoloSpeak"
      >
        <div
          class="d-flex justify-content-center align-items-center my-9 mt-md-12 mb-md-10 text-center display-1"
        >
          बल्कि मजबूरी थी 20
        </div>
        <div
          class="d-flex justify-content-center align-items-center my-9 mt-md-12 mb-md-10 text-center display-1"
        >
          <canvas
            id="visualizer"
          />
        </div>
        <div
          class="mt-12 mt-md-14"
        >
          <div
            class="d-flex flex-column flex-md-row justify-content-md-center align-items-center"
          >
            <button
              class="mx-md-6 order-2 order-md-1 my-2 my-md-0 light secondary btn btn-light"
              data-testid="Button"
              type="button"
            >
              startRecording
            </button>
            <button
              class="mx-md-6 order-3 order-md-3 my-2 my-md-0 info primary btn btn-primary"
              data-testid="Button"
              disabled=""
              type="button"
            >
              submit
            </button>
          </div>
          <div
            class="d-flex justify-content-center mt-4 mt-md-13"
          >
            <button
              class="transparent tertiary btn btn-transparent"
              data-testid="Button"
              type="button"
            >
              skip
            </button>
          </div>
        </div>
        <div
          class="d-flex align-items-center mt-10 mt-md-14"
        >
          <div
            class="flex-grow-1"
          >
            <div
              class="progress progress"
            >
              <div
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="20"
                class="progress-bar bg-primary"
                role="progressbar"
                style="width: 20%;"
              />
            </div>
          </div>
          <span
            class="ms-5"
          >
            1/5
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
